### ISSUE TITLE
False positives from tests/test_self.py

### ISSUE THREAD

**Issue by greg-ww** [pylint-dev/pylint#10450]
I noticed that some tests in test_self.py fail to validate the behavior they are meant to verify due to the way with ```pytest.raises(SystemExit):``` is being used blocking the asserts from ever being run. 

For example in tests/test_self.py::TestCallbackOptions::test_errors_only you can change 
```@staticmethod
    def test_errors_only() -> None:
        """Test the --errors-only flag."""
        with pytest.raises(SystemExit):
            run = Run(["--errors-only"])
            assert run.linter._error_mode
```
to
```@staticmethod
    def test_errors_only() -> None:
        """Test the --errors-only flag."""
        with pytest.raises(SystemExit):
            run = Run(["--an-invalid-option"])
            assert run.linter._error_mode
```
and it still passes perfectly happily but when you look at the output there were errors

```
$ pytest tests/test_self.py::TestCallbackOptions::test_errors_only -s
============================================================= test session starts =============================================================
platform win32 -- Python 3.12.3, pytest-8.3.5, pluggy-1.5.0
rootdir: D:\Repos\pylint
configfile: pyproject.toml
plugins: anyio-4.9.0, langsmith-0.4.1
collected 1 item                                                                                                                               

tests\test_self.py usage: pylint [options]
pylint: error: Unrecognized option found: an-invalid-option
.

============================================================== 1 passed in 0.12s ============================================================== 
```

You can even just change the assert line to ```assert False``` and it will still pass every time. 

This affects multiple tests in test_self.py I confirmed this behavior in test_errorrs_only and test_verbose.

**Comment by Pierre-Sassoulas** [comment]
Thank you, nice catch !

**Comment by Ayushikotiyal** [comment]
Hi @Pierre-Sassoulas ,  

I’m new to open source and would love to start contributing.  
This issue looks like a good first step for me — may I work on it?  
If it’s available, could you please assign it to me?  

Thank you!

**Comment by Pierre-Sassoulas** [comment]
Hey @Ayushikotiyal thank you for your interest in pylint ! I've assigned you to the issue :)

**Comment by Ayushikotiyal** [comment]
Hi @Pierre-Sassoulas
Raised  a PR PTAL.
